### PR TITLE
httpbakery: Add context to to client methods.

### PR DESCRIPTION
**Pull request by mhilton**
Add a DoWithContext method and use the context with all HTTP
requests. Also add context to DischargeAll and HandleError methods.

### CODE REVIEW

**Review by mattyw — APPROVED** [review]

File: httpbakery/client.go [hunk 1]
```diff
@@ -198,20 +180,30 @@ func (c *Client) dischargeAcquirer() DischargeAcquirer {
...
+// If the request fails with a discharge-required error, any required
+// discharge macaroons will be acquired, and the request will be repeated
+// with those attached.
+//
+// If the required discharges were refused by a third party, an error
+// with a *DischargeError cause will be returned.
+//
+// If interaction is required by the user, the client's WebPageVisitor,
+// or VisitWebPage function will be used to perform interaction. An error
+// with a *InteractionError cause will be returnedif this interaction
```
Comment (rogpeppe) [inline httpbakery/client.go]: returned if

File: httpbakery/client.go [hunk 2]
```diff
@@ -518,6 +510,22 @@ func (c *Client) interact(location, visitURLStr, waitURLStr string) (*macaroon.M
 	return resp.Macaroon, nil
 }
 
+// relativeURL returns newPath relative to an original URL.
```
Comment (rogpeppe) [inline httpbakery/client.go:513]: Thanks for moving this.

File: httpbakery/client_test.go [hunk 3]
```diff
@@ -238,7 +238,7 @@ func assertDischargeServerDischargesConditionForVersion(c *gc.C, cond string, ve
 	}, nil)
 	c.Assert(err, gc.IsNil)
```
Comment (rogpeppe) [inline httpbakery/client_test.go:239]: I wonder if we should test whether the context actually gets through to the parts we intend it to.

File: httpbakery/client.go [hunk 4]
```diff
@@ -198,20 +180,30 @@ func (c *Client) dischargeAcquirer() DischargeAcquirer {
...
+// with a *DischargeError cause will be returned.
+//
+// If interaction is required by the user, the client's WebPageVisitor,
+// or VisitWebPage function will be used to perform interaction. An error
+// with a *InteractionError cause will be returnedif this interaction
+// fails. See WebBrowserVisitor for a possible implementation of
+// WebPageVisitor.
+//
+// DoWithContext may add headers to req.Header.
+func (c *Client) DoWithContext(ctx context.Context, req *http.Request) (*http.Response, error) {
```
Comment (rogpeppe) [inline httpbakery/client.go:205]: I'm not sure about having another Do variant again now we've eliminated DoWithBody.. For example, what about DoWithContextAndCustomError?

How about just adding a context to Do and DoWithCustomError and don't bother to derive the context from the request?
At some point when we can use Go 1.7 we can break the API and take the context argument out and derive it from the request.

For now, I think all our pre-1.7 programs will need to pass in the request explicitly anyway.

**Review by rogpeppe — CHANGES_REQUESTED** [review]
LGTM but I have a concern about the proliferation of Do methods.

File: httpbakery/client.go (hunk 1 above)
Comment (mhilton) [reply httpbakery/client.go]: done

File: httpbakery/client.go (hunk 4 above)
Comment (mhilton) [reply httpbakery/client.go:205]: As discussed this is the interface we are using with httprequest

File: httpbakery/client.go (hunk 2 above)
Comment (mhilton) [reply httpbakery/client.go:513]: you're welcome

**Review by rogpeppe — APPROVED** [review]

**Pull request merged**